simhash/__init__.py

```diff
@@ -64,36 +64,68 @@ def _tokenize(self, content):
         ans = self._slide(content)
         return ans
 
-    def build_by_text(self, content):
-        features = self._tokenize(content)
-        features = {k:sum(1 for _ in g) for k, g in groupby(sorted(features))}
-        return self.build_by_features(features)
-
-    def build_by_features(self, features):
+    def update(self, features):
         """
         `features` might be a list of unweighted tokens (a weight of 1
-                   will be assumed), a list of (token, weight) tuples or
-                   a token -> weight dict.
+                   will be assumed), a list of (token, weight) tuples, a
+                   token -> weight dict or a string.
         """
-        v = [0] * self.f
-        masks = [1 << i for i in range(self.f)]
+        if isinstance(features, (str, unicode)):
+            # If we receive a string, we will prepend our buffer, tokenize the
+            # result, and withhold the last token in our buffer.
+            if self.buffer:
+                features = self.buffer + features
+            if not features:
+                return
+            features = self._tokenize(features)
+            self.buffer = features.pop()
+            features = {k:sum(1 for _ in g) for k, g in groupby(sorted(features))}
         if isinstance(features, dict):
             features = features.items()
         for f in features:
             if isinstance(f, basestring):
-                h = self.hashfunc(f.encode('utf-8'))
+                h = f.encode('utf-8')
                 w = 1
             else:
                 assert isinstance(f, collections.Iterable)
-                h = self.hashfunc(f[0].encode('utf-8'))
+                h = f[0].encode('utf-8')
                 w = f[1]
+            self.features[h] += w
+
+    # Preserve old interface.
+    build_by_features = update
+    build_by_text = update
+
+    def finalize(self):
```

Comment:
Why do we need `finalize`? Can we finalize internally when we get the value so that user don't need to call this function?